### DISCUSSION TITLE
how to post image avoiding modification to keep it AS-IS (original)?

### DISCUSSION THREAD

**Discussion by itaru2622** [bluesky-social/atproto#1990] [How to do it?] [answered]
I'm trying to post and share image AS-IS  but no success yet.

even I post small jpeg file,  it is applied modification ( cannot keep original).
my conditions are:
- client to post image: https://github.com/bluesky-social/cookbook/tree/main/python-bsky-post
- server:                          official bsky.social
- client to view and download image: official web client ( bsky.app )

post with below image (512x512 24kB) becomes bigger (2000x2000 64kB)  when downloading after posting.
![icon2_noexif](https://github.com/bluesky-social/atproto/assets/70509350/28aea9c5-a468-4d70-921a-adf767f7589d)

I found some condition on  posting image, such as https://github.com/bluesky-social/atproto/pull/442/ .
but couldn't found how to post image  to avoid modification (keep it original).

cf. above test image created from current social-app icon.

**Comment by DavidBuchanan314** [comment, marked as answer]
Images do get posted as-is, but they don't get served back as-is by the default bsky.app client. You can request the originals via the `com.atproto.sync.getBlob` API

**Reply by itaru2622** [reply]
@DavidBuchanan314 thank you for your response. I will test getBlob API. Could you let me know (point out) where client modifies image on serving?

**Reply by bnewbold** [reply]
@itaru2622 to add more context to David's answer:

Clients upload blobs to the logged-in account's PDS. Clients *may* (and *should* in most cases) do thing like remove image EXIF metadata, as a privacy concern. But generally the original blob file is stored as-is in the PDS. You can fetch this original blob using `com.atproto.sync.getBlob`, but note that you need to make this request to the relevant PDS.

Separately, the bsky appview (which is the network service which implements the bluesky micro-blogging application) crawls images from all the PDS instances and puts them in a CDN. The API responses from the appview include URLs pointing to the CDN, not to the original PDS instances. This ensures consistent latency, and puts most of bandwidth burden on the appview instead of PDS instances. This architectural decision, and the details of which CDN to use, how images are resized, etc, are an internal implementation detail of the appview, not something dictated by atproto (the underlying protocol).

To be transparent about it, currently bluesky (the company) uses an free software daemon called imgproxy (https://imgproxy.net/) to do the resizing, and our CDN service provider is https://bunny.net/.

**Reply by itaru2622** [reply]
@bnewbold thank you for your detail description.

I'd like to confirm my understanding:
as far as using current official client (such as social-app),
there is no method to view original image  stored into PDS, because client serves image via CDN with some modifying for latency.

**Reply by bnewbold** [reply]
That's correct.

You could file a feature request (github issue) over in the https://github.com/bluesky-social/social-app repo if you'd like the official app to have some way to download the original image (eg, via a drop-down menu or something like that). I'm not sure that would be prioritized, but seems like a reasonable request.

**Reply by itaru2622** [reply]
@bnewbold thank you for your comment. I posted this as issue, https://github.com/bluesky-social/social-app/issues/2380

**Reply by ProgrammerWhoPrograms** [reply]
>the bsky appview (which is the network service which implements the bluesky micro-blogging application) 

Are alternative app views for bluesky or other social modes possible?

**Reply by itaru2622** [reply]
@ProgrammerWhoPrograms I couldn't understand your question. could you describe what you concern in more specific? because bnewbold described lots of things in the above comment.

**Comment by itaru2622** [comment]
@bnewbold @DavidBuchanan314 thank you for your responses, I could download original image with com.atproto.sync.getBlob after getting cid within app.bsky.feed.getAuthorFeed

**Comment by itaru2622** [comment]
I confirmed that patch to below line resolve this issue, through debugging.

https://github.com/bluesky-social/atproto/blob/e4ec7af03608949fc3b00a845f547a77599b5ad0/packages/bsky/src/services/feed/views.ts#L369-L373

by change URL to getBlob endpoint like below:
```javascript
fullsize: `https://bsky.social/xrpc/${ids.ComAtprotoSyncGetBlob}?did=${did}&cid=${img.image.ref}`,
```

**Comment by syui** [comment]
@itaru2622 Is there a way to display images on the timeline in self-hosting?

It seems that the image resizing service is set to `BSKY_IMG_URI_ENDPOINT`, but I don't know if such a service is running in the environment.

Currently, access the image using the following URI:

```sh
1. https://appview.${host}/img/avatar/plain/${did}/${cid}@jpeg

2. https://${host}/xrpc/com.atproto.sync.getBlob?did=${did}&cid=${cid}
```


[google translate]

**Reply by syui** [reply]
thx, it's probably working fine with this patch.

https://github.com/itaru2622/bluesky-selfhost-env/blob/master/patching/160-social-app-disable-hackModifyThumbnailPath.diff

**Reply by itaru2622** [reply]
@syui 
try below env var instead of BSKY_IMG_URI_ENDPOINT.

https://github.com/bluesky-social/atproto/blob/60df3fc652b00cdff71dd9235d98a7a4bb828f05/services/bsky/README.md?plain=1#L16

BSKY_IMG_URI_ENDPOINT is different matter than this discussion  topic.
it is better to create new one.
